### PR TITLE
fix: render slashes in assistant messages instead of &#47; (#549)

### LINKED ISSUE #549: [BUG] slashes not rendered
### App Version

3.4.2

### Platform

- [ ] Android
- [x] iOS

### How you installed it

App Store

### What happened?

Slashes don’t get printed. Instead it just shows &#47; .

### What should have happened?

Print /

### How to reproduce

Let the assistant send a message with an link for example (https://) 

### Does this happen every time?

Yes, every time

### When did this start?

I think with the last update.

### Screenshots

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by cogwheel0**
## Summary

Fixes #549 — assistant messages rendered `/` as the literal HTML entity `&#47;` (e.g. `https:&#47;&#47;`), starting with 3.4.1/3.4.2.

**Root cause:** `semantic_message_builder.dart` reconstructs assistant content (answer text + reasoning/tool `<details>` blocks) and escaped text with Dart's **default** `HtmlEscape()` — `HtmlEscapeMode.unknown`, which encodes `/`→`&#47;` and `'`→`&#39;` in addition to `& < > "`. The markdown renderer decodes those entities in prose (via the `markdown` package's `DecodeHtmlSyntax`), but **never inside inline code, fenced code blocks, or link destinations** — so slashes there (URLs, paths, `a/b`) leaked to the screen. OpenWebUI's server uses Python `html.escape`, which does not escape `/`, so this over-escaping was Conduit-specific.

## Changes

- Switch the shared escaper to `HtmlEscapeMode.attribute` (`& < > "` only) — drops the leaky `/` and `'` while keeping everything needed to neutralize HTML tags and to keep double-quoted `<details>` attributes intact.
- Add `_escapeText`: plain answer text (`SemanticTextBlock`) is now escaped **only outside** code spans/fenced blocks, so `<`, `>`, `&`, `"` no longer leak inside code either (the same failure class, extended). `<details>` attributes/summaries/bodies keep full escaping — they're HTML-unescaped wholesale by `DetailsBlockSyntax._decode` at parse time and don't leak, and the body must keep escaping `</details>` for tag-counting.

## Safety

The one real risk was UI-spoofing: skipping escaping inside "code" could let a model inject a fake `<details type="reasoning">` block. Guard: inline-code matching is restricted to a **single line** (`` `[^`\n]+?` ``), a strict subset of real code spans, so a multi-line span can't hide a line-leading tag from the block parser. Validated adversarially against the real `DetailsBlockSyntax` — inline, fenced, multi-line-inline, double-backtick, and indented `<details>` spoofs all stay neutralized.

## Tests

- New `semantic_message_builder_test.dart` (8 cases): slashes, apostrophes, inline-code, fenced-code, tag neutralization, multi-line-inline injection — **8/8**.
- `openwebui_stream_parser_test.dart` — **36/36** (updated one assertion that hard-coded the `/`-escaping byproduct: `&lt;&#47;details&gt;` → `&lt;/details&gt;`).
- `conversation_parsing_test.dart` — **44/44** (incl. `<details>`-neutralization tests).
- `flutter analyze` — clean.

> [!NOTE]
> ### Fix slash rendering in assistant messages by switching HTML escape mode
> - Changes `_semanticHtmlEscape` from `HtmlEscape()` to `HtmlEscape(HtmlEscapeMode.attribute)` so `/` and apostrophes are no longer converted to `/` and `'` in assistant message output.
> - Adds `_escapeText`, which escapes HTML-significant characters only in prose regions, leaving fenced code blocks (backtick and tilde) and inline code spans untouched.
> - Updates `renderSemanticMessageBlocks` to use `_escapeText` instead of the raw `_escape` call for `SemanticTextBlock` content.
> - Behavioral Change: `</details>` and similar HTML tags in plain text are now neutralized via escaping, while the same tags inside code spans/blocks are emitted verbatim.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 76e787f.</sup>
> 
>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved HTML escaping for rendered semantic messages to remove unintended extra entities and better neutralize tag spoofing (including `</details>`/`<summary>`-style sequences).
  * Prose is now escaped while preserving URLs (e.g., forward slashes) and apostrophes; inline/backtick and fenced code blocks remain unmodified.

* **Tests**
  * Updated the stream parser test to expect `</details>` to render as `&lt;/details&gt;`.
  * Added a new regression suite covering escaping and code-region boundary/spoofing edge cases.

<h3>Greptile Summary</h3>

This PR fixes assistant message rendering so slash and apostrophe characters no longer appear as HTML entities. The main changes are:

- Switch semantic HTML escaping to attribute mode for generated message markup.
- Escape plain answer text while preserving inline code and fenced code regions.
- Keep generated `<details>` attributes, summaries, and bodies escaped for tag safety.
- Add tests for slash rendering, code spans, fenced code, CRLF input, and details-block spoofing cases.

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge with minimal risk.

The change is localized to semantic message serialization and tests. The implementation keeps generated details markup escaped while preserving markdown code regions. The tests cover the main rendering and spoofing edge cases reviewed.

No files require special attention.
</details>

<details><summary><h3><a href="https://www.greptile.com/trex">T-Rex</a> T-Rex Logs</h3></summary>

**What T-Rex did**
- We confirmed that Flutter and Dart are not on PATH, based on the semantic-message-availability.log.
- The semantic message builder test invocation could not start because Flutter was not found on PATH, with /bin/sh: flutter: not found and exit code 127, as shown in the semantic-message-builder-test-blocker.log.

<a href="https://app.greptile.com/trex/runs/13892575/artifacts"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifactsDark.svg?v=4"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifacts.svg?v=4">View all artifacts</picture></a>

<sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| lib/core/services/semantic_message_builder.dart | Switches semantic HTML escaping to attribute mode and adds line-aware text escaping that preserves inline and fenced code while escaping prose and generated details content. |
| test/core/services/openwebui_stream_parser_test.dart | Updates the escaping expectation for generated details content now that closing slashes are no longer over-escaped. |
| test/core/services/semantic_message_builder_test.dart | Adds tests for entity rendering, code preservation, tag neutralization, CRLF handling, and prior spoofing edge cases. |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
participant Parser as OpenWebUI stream parser
participant Builder as renderSemanticMessageBlocks
participant Escaper as _escapeText / _escape
participant Markdown as Markdown renderer

Parser->>Builder: SemanticTextBlock + SemanticDetailsBlock
Builder->>Escaper: Escape text blocks line-by-line
Escaper-->>Builder: Prose escaped, inline/fenced code preserved
Builder->>Escaper: Escape details attrs/summary/body
Escaper-->>Builder: Details content tag-safe
Builder->>Markdown: Serialized markdown/details markup
Markdown-->>Markdown: Decode prose/entities, preserve code text
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
participant Parser as OpenWebUI stream parser
participant Builder as renderSemanticMessageBlocks
participant Escaper as _escapeText / _escape
participant Markdown as Markdown renderer

Parser->>Builder: SemanticTextBlock + SemanticDetailsBlock
Builder->>Escaper: Escape text blocks line-by-line
Escaper-->>Builder: Prose escaped, inline/fenced code preserved
Builder->>Escaper: Escape details attrs/summary/body
Escaper-->>Builder: Details content tag-safe
Builder->>Markdown: Serialized markdown/details markup
Markdown-->>Markdown: Decode prose/entities, preserve code text
```

</a>
</details>

<sub>Reviews (6): Last reviewed commit: ["fix: normalize CRLF line endings in \_esc..."](https://github.com/cogwheel0/conduit/commit/76e787fab3c410b8c1309a62ba78daedd68153a4) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=43058852)</sub>

### CODE REVIEW

File: lib/core/services/semantic_message_builder.dart [hunk 1]
```diff
@@ -184,3 +190,26 @@ String _markdownCodeFence(String code, String language) {
 }
 
 String _escape(String value) => _semanticHtmlEscape.convert(value);
+
+// Matches fenced code blocks (```...```) and single-line inline code spans.
+// Inline spans are intentionally matched one line at a time: a multi-line span
+// must never be able to hide a line-leading `<details>`/`<summary>` from the
+// block parser (see [_escapeText]).
+final _codeRegionPattern = RegExp(r'```[\s\S]*?```|`[^`\n]+?`');
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/core/services/semantic_message_builder.dart]: Fixed in 0f253504. `_codeRegionPattern` now matches `~~~` fences as well as ```` ``` ````.

While verifying, the naive fix (an unanchored `~~~[\s\S]*?~~~` alternative) revealed a pre-existing hole: an unanchored fence alternative can pair a **mid-line** fence marker with a later fence around a line-leading `<details>`, skipping its escaping so it renders as a spoofed reasoning block. So the fences are now **anchored to the start of a line** (0-3 space indent, backreferenced fence run) for both ```` ``` ```` and `~~~`. Matched regions are a strict subset of what the markdown block parser treats as code, so escaping is only skipped where the parser would not begin a block. Mismatched/unclosed fences fall back to being escaped as prose (safe direction).

Added regression tests for `~~~` fences and the mid-line-fence injection case.

[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/core/services/semantic_message_builder.dart]: Fixed in 0f253504. `_codeRegionPattern` now matches `~~~` fences as well as ```` ``` ````.

While verifying, the naive fix (an unanchored `~~~[\s\S]*?~~~` alternative) revealed a pre-existing hole: an unanchored fence alternative can pair a **mid-line** fence marker with a later fence around a line-leading `<details>`, skipping its escaping so it renders as a spoofed reasoning block. So the fences are now **anchored to the start of a line** (0-3 space indent, backreferenced fence run) for both ```` ``` ```` and `~~~`. Matched regions are a strict subset of what the markdown block parser treats as code, so escaping is only skipped where the parser would not begin a block. Mismatched/unclosed fences fall back to being escaped as prose (safe direction).

Added regression tests for `~~~` fences and the mid-line-fence injection case.

File: lib/core/services/semantic_message_builder.dart [hunk 2]
```diff
@@ -184,3 +190,26 @@ String _markdownCodeFence(String code, String language) {
...
+/// `https:&#47;&#47;`). Text outside code is still escaped so a model cannot
+/// emit a literal `<details>`/`<summary>` block that renders as a spoofed
+/// reasoning/tool section. Unlike [_escape], this is only safe for top-level
+/// answer text; `<details>` attributes/summaries/bodies are HTML-unescaped
+/// wholesale at parse time and must keep full escaping.
+String _escapeText(String value) => value.splitMapJoin(
+  _codeRegionPattern,
+  onMatch: (match) => match[0] ?? '',
+  onNonMatch: _escape,
+);
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/core/services/semantic_message_builder.dart]: Fixed in 0f253504. `_codeRegionPattern` now matches `~~~` fences as well as ```` ``` ````.

While verifying, the naive fix (an unanchored `~~~[\s\S]*?~~~` alternative) revealed a pre-existing hole: an unanchored fence alternative can pair a **mid-line** fence marker with a later fence around a line-leading `<details>`, skipping its escaping so it renders as a spoofed reasoning block. So the fences are now **anchored to the start of a line** (0-3 space indent, backreferenced fence run) for both ```` ``` ```` and `~~~`. Matched regions are a strict subset of what the markdown block parser treats as code, so escaping is only skipped where the parser would not begin a block. Mismatched/unclosed fences fall back to being escaped as prose (safe direction).

Added regression tests for `~~~` fences and the mid-line-fence injection case.

File: lib/core/services/semantic_message_builder.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/core/services/semantic_message_builder.dart]: Follow-up in 29913862: the **unclosed trailing fence** case you flagged is now handled too. Added a second, line-anchored alternative to `_codeRegionPattern` that matches an opening ```` ``` ````/`~~~` fence running to end of input — mirroring how both the block parser and `normalize()` treat an unclosed fence (as a code block to EOF). Contents of unclosed fences are now preserved instead of escaped as prose, and because the opener stays anchored to a line start, a `<details>` after an unclosed fence remains inside the code block rather than rendering as a spoofed section. Regression tests added for both.

File: lib/core/services/semantic_message_builder.dart [hunk 3]
```diff
@@ -184,3 +190,39 @@ String _markdownCodeFence(String code, String language) {
...
+// [ConduitMarkdownPreprocessor.normalize] treat it — e.g. mid-stream, or when a
+// model forgets the closing fence), and (3) a single-line inline span. Because
+// every fence is anchored to a line start, the matched regions are a strict
+// subset of what the markdown block parser treats as code, so escaping is only
+// ever SKIPPED where the parser would not begin a block — a line-leading
+// `<details>`/`<summary>` can never be smuggled past [_escapeText] via a
+// mid-line fence marker. Inline code is matched one line at a time for the same
+// reason.
+final _codeRegionPattern = RegExp(
+  r'(?:^|\n)[ ]{0,3}(`{3,}|~{3,})[^\n]*\n[\s\S]*?\n[ ]{0,3}\1[ \t]*(?=\n|$)'
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/core/services/semantic_message_builder.dart]: Good catch — real hole, fixed in db4d6c44. The closing-fence branch now allows the closing line to repeat the fence character beyond the opening length (`\1[`~]*`), matching CommonMark's "closing fence may be longer" rule. Verified against your PoC (open ```` ``` ````, close ```` ```` ````, then a line-leading `<details>`): the block now closes at the longer fence, so the following `<details>` is escaped as prose instead of falling through to the unclosed-to-EOF branch. Over-matching a close only ever escapes *more* (a harmless cosmetic entity), never less, so it can't reintroduce injection. Regression test added.

File: lib/core/services/semantic_message_builder.dart [hunk 4]
```diff
@@ -184,3 +190,39 @@ String _markdownCodeFence(String code, String language) {
...
+  r'|(?:^|\n)[ ]{0,3}(?:`{3,}|~{3,})[^\n]*\n[\s\S]*$'
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/core/services/semantic_message_builder.dart]: Good catch — real hole, fixed in db4d6c44. The closing-fence branch now allows the closing line to repeat the fence character beyond the opening length (`\1[`~]*`), matching CommonMark's "closing fence may be longer" rule. Verified against your PoC (open ```` ``` ````, close ```` ```` ````, then a line-leading `<details>`): the block now closes at the longer fence, so the following `<details>` is escaped as prose instead of falling through to the unclosed-to-EOF branch. Over-matching a close only ever escapes *more* (a harmless cosmetic entity), never less, so it can't reintroduce injection. Regression test added.

File: lib/core/services/semantic_message_builder.dart [hunk 5]
```diff
@@ -184,3 +190,40 @@ String _markdownCodeFence(String code, String language) {
...
+// block parser treats as code, so escaping is only ever SKIPPED where the
+// parser would not begin a block — a line-leading `<details>`/`<summary>` can
+// never be smuggled past [_escapeText] via a mid-line fence marker or a longer
+// closing fence. Inline code is matched one line at a time for the same reason.
+final _codeRegionPattern = RegExp(
+  r'(?:^|\n)[ ]{0,3}(`{3,}|~{3,})[^\n]*\n[\s\S]*?\n[ ]{0,3}\1[`~]*[ \t]*(?=\n|$)'
+  r'|(?:^|\n)[ ]{0,3}(?:`{3,}|~{3,})[^\n]*\n[\s\S]*$'
+  r'|`[^`\n]+?`',
+  multiLine: true,
+);
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/core/services/semantic_message_builder.dart]: Fixed in 4160d81c — and thanks, this independently confirms the same hole. I replaced the fence **regex** entirely with a line-based scanner that opens a backtick fence only when the info string has no backtick (`^ {0,3}(`{3,})[^`]*$`), exactly as you suggest, and opens a tilde fence on `^ {0,3}(~{3,})`. Since open/close detection now mirrors the parser, a line like `` ```x`y `` is treated as prose and the following `<details>` is escaped. Regression test added (`escapes a tag after a non-fence line with backtick in info`). This thread references the superseded regex, so resolving as outdated.

File: lib/core/services/semantic_message_builder.dart [hunk 6]
```diff
@@ -184,3 +190,80 @@ String _markdownCodeFence(String code, String language) {
...
+/// including unclosed fences (which extend to end of input, e.g. mid-stream),
+/// so the skipped regions are exactly the parser's code and escaping is never
+/// skipped where the parser would begin a block. Inline code is matched a
+/// single line at a time so a multi-line span cannot hide a line-leading tag.
+///
+/// Unlike [_escape], this is only safe for top-level answer text; `<details>`
+/// attributes/summaries/bodies are HTML-unescaped wholesale at parse time and
+/// must keep full escaping.
+String _escapeText(String value) {
+  final lines = value.split('\n');
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/core/services/semantic_message_builder.dart:227]: Fixed in 76e787fa. `_escapeText` now strips a trailing `\r` from each line before the fence scan, so CRLF `~~~` (and ```` ``` ````) blocks are recognized as code and their contents are preserved. Confirmed via a CRLF repro that the `~~~` case leaked before and no longer does; also dropped the now-redundant `\r?` from `_closingFence`. Regression test added (`recognizes fenced code blocks with CRLF line endings`).
